FastAPI: migrate history contents routes

davelopez opened the pull request:
This is a follow up on #12231 and part of #10889

This PR includes the new `FastAPIHistoryContents` controller that replaces the old legacy routes for galaxy/api/history_contents.py and some additional refactorings, documentation, and types annotations.

## Additional comments / Issues
- I took the liberty of *temporarily* marking some routes and parameters as deprecated as some of them looked duplicated or legacy. This should only affect how they are displayed in the interactive documentation and can be reverted easily if they are incorrectly marked.

  ![Screenshot from 2021-10-11 20-19-06](https://user-images.githubusercontent.com/46503462/136836478-3ce5bf28-64d1-46b3-a07b-9571de1df1fc.png)

  ![Screenshot from 2021-10-11 20-22-29](https://user-images.githubusercontent.com/46503462/136836704-58eb741e-9b8b-433b-8132-a584416a1ede.png)

- I had to add in some cases several helper functions to use as FastAPI dependencies for retrieving the query parameters and parse them in a custom way. There was no way of directly using a pydantic model in those cases because the documentation was not rendered correctly.
- Temporarily adding [here the '/api/dataset_collection/{id}/download' route](https://github.com/galaxyproject/galaxy/pull/12578/commits/b7c5fc8e0dc3d8a8b3c06677e5a3c59a72fb7442#diff-d7789a8b5b8606bd040d59409203acc178c14fe4ec4d6b38e8955d87dc4ecea1R414-R419) as both endpoints were redirected to the same operation. It can be moved to the dataset_collections API and reuse this service later.
- I managed to make the `StreamingResponse` here work with minimal changes in ZipstreamWrapper in various endpoints, but I'm not sure if this is the best solution or we need to think of a different solution when the legacy controllers are removed. Not that I have a particular concern, just I don't know very well how to deal with files efficiently in the backend yet.
- I had to mark a bunch of API tests with `json=True`. It should be a temporal change until we can make JSON the default, see https://github.com/galaxyproject/galaxy/pull/12152#issuecomment-864134487.
- The `contents_near` endpoint was [exposed in *raw* mode](https://github.com/galaxyproject/galaxy/pull/12578/commits/1fff6648d7717d97ff369fc3374d761098cb5e19#diff-d7789a8b5b8606bd040d59409203acc178c14fe4ec4d6b38e8955d87dc4ecea1R1059). I changed it back to the normal JSON default, I don't know if there was a particular reason for the change. Also, this endpoint treats the filtering parameters in a different way, so instead of reusing the existing `FilterQueryParams` with *q*/*qv* pairs, it uses arbitrary query parameters. I managed to make it work in FastAPI mode by directly parsing the query parameters from the request, but unfortunately, this makes the filtering parameters not visible in the interactive documentation, although I guess this endpoint is quite for *internal use* so it may be ok.


## How to test the changes?
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - Go to the interactive API documentation at http://127.0.0.1:8080/api/docs#/histories

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

davelopez committed:
Remove unused dependencies

Leftovers from previous refactoring.

davelopez committed:
Fix some issues in HDADetailed schema model

- Add some missing properties for metadata
- Make some properties optional

davelopez committed:
Refactor HistoryContentsService

- Simplify models and move API documentation to dependency functions. This allows to properly display the documentation in OpenAPI when there are multiple models.

davelopez committed:
Allow custom model properties to be serialized

davelopez committed:
Add FastAPI route for index operation

The query parameters are retrieved by dependency functions to be able to render the documentation correctly. Directly using the models as dependecies will work but the documentation is not shown in OpenAPI for some reason.

davelopez committed:
Add FastAPI route for show operation

davelopez committed:
Small refactoring + docstrings

davelopez committed:
Add FastAPI route for `index_jobs_summary` operation

davelopez committed:
Add FastAPI route for `show_jobs_summary` operation

davelopez committed:
Refactor `download_dataset_collection`

- Rename method and return ZipstreamWrapper to adapt response depending on API framework (legacy or FastAPI)
- Move response related code from service to controller
- Raise proper exception and do not expose internal exception on API response

davelopez committed:
Add FastAPI route for `download_dataset_collection`

- Temporarily adding here the '/api/dataset_collection/{id}/download' route. It can be moved to the dataset_collections API and reuse this service later.
- I managed to make the StreamingResponse here work with minimal changes in ZipstreamWrapper, but I'm not sure if this is the best solution or we need to think of a different solution when the legacy controllers are removed.

davelopez committed:
Adapt history contents API tests to use JSON payload

See https://github.com/galaxyproject/galaxy/pull/12152 for details

davelopez committed:
Fix `create` endpoint and related models

- Make some fields optional as they depend on the type of request
- Explicitly get history content type from payload or path param

davelopez committed:
Add FastAPI route for `create` operation

- Also marks `/api/histories/{history_id}/contents` as deprecated because explicit sounds better than implicit, but we can always remove the deprecation flag if that's ok.

davelopez committed:
Adapt more API tests around history contents to use JSON payload

See https://github.com/galaxyproject/galaxy/pull/12152 for details

davelopez committed:
Add support for nested collections in `CreateNewCollectionPayload`

davelopez committed:
Add FastAPI route for `update_permissions` operation

- Make `set_permissions` the default action in payload

davelopez committed:
Simplify UpdateHistoryContentsBatchPayload model

Also provide example in the schema.

davelopez committed:
Add FastAPI route for `update_batch` operation

davelopez committed:
Unify put and delete requests in some API tests

Replace raw requests by custom `_put` and `_delete`

davelopez committed:
Unify content type query parameter handling

Use `dataset` as default value and reuse schema definition.

davelopez committed:
Add FastAPI route for `update` operation

davelopez committed:
Adapt more API tests around history contents to use JSON payload

davelopez committed:
Fix dataset_details not being populated in `dev` version index operation

Sometimes the endpoint get called with the parameter alias `details` instead of `dataset_details`.

davelopez committed:
Add FastAPI route for `validate` operation

davelopez committed:
Add FastAPI route for `delete` operation

- Refactor payload and response model

davelopez committed:
Move direct request access outside of service function in `archive` operation

This will allow to reuse the archive service function in FastAPI mode too. Also add model for dry run result.

davelopez committed:
Add FastAPI route for `archive` operation

davelopez committed:
Refactor `contents_near` operation

- Move direct response manipulation from service to endpoint (FastAPI compatibility)
- Add models for stats and results
- Expose legacy endpoint as JSON (not raw)

davelopez committed:
Add FastAPI route for `contents_near` operation

This endpoint handles the filter query parameters differently and the only way I  could make it work in FastAPI mode is by directly parsing the query parameters from the requests. Since the name of the parameters can be anything there is no way of displaying this  information in OpenAPI.

davelopez committed:
Fix mypy

I wonder why `tox -e mypy` does not catch this... maybe uses a different version?

davelopez committed:
Exclude known query params from `contents_near` filter params

Since we are directly accessing the request's query_params we also need to exclude the known params that are already parsed by FastAPI or they may be treated as filter params too.

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -628,45 +634,55 @@ class HDCADetailed(HDCASummary):
 
-    items: List[Union[UpdateHDAPayload, UpdateHDCAPayload]] = Field(
+class UpdateHistoryContentsBatchPayload(HistoryBase):
+    """Contains property values that will be updated for all the history`items` provided."""
```
```suggestion
    """Contains property values that will be updated for all the history `items` provided."""
```

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -2353,29 +2369,108 @@ class UpdateDatasetPermissionsPayload(Model):
+    purge: bool = Field(
+        default=False,
+        title="Purge",
+        description="Whether to remove from disk the target HDA or child HDAs of the target HDCA.",
```
```suggestion
        description="Whether to remove the dataset from storage. Datasets will only be removed from storage once all HDAs or LDDAs that refer to this datasets are deleted.",
```

mvdbeek commented on lib/galaxy/schema/schema.py:
```diff
@@ -2353,29 +2369,108 @@ class UpdateDatasetPermissionsPayload(Model):
-AnyHDA = Union[HDASummary, HDADetailed, HDABeta]
-AnyHistoryContentItem = Union[AnyHDA, HDCASummary, HDCADetailed, HDCABeta]
+class HistoryContentsArchiveDryRunResult(BaseModel):
+    """The structure of the archive for debugging.
```
Do you want to update the docstring here ?

davelopez replied:
Something like this? https://github.com/galaxyproject/galaxy/pull/12578/commits/42d7b3f64d7cea91887d6b9cfd33c043bfc529ba

mvdbeek replied:
Perfect!

mvdbeek commented on lib/galaxy/webapps/galaxy/api/history_contents.py:
```diff
@@ -34,20 +54,673 @@
+        while maintaining approximate collection structure.
+        """
+        archive = self.service.get_dataset_collection_archive_for_download(trans, id)
+        return StreamingResponse(archive.get_iterator(), headers=archive.get_headers())
```
Can we return the ZipstreamWrapper.response() contents here if mod_zip is activated? This is probably too much to handle within the web process.

davelopez replied:
Oh! You are absolutely right I totally overlooked this option. Thanks for the catch!

davelopez replied:
Ok, with https://github.com/galaxyproject/galaxy/pull/12578/commits/3f5722624c62715f77e907f6f46fac49e15da549, a quick test with `upstream_mod_zip: true` returns this kind of response:

```
HTTP/1.1 200 OK
date: Thu, 14 Oct 2021 08:47:43 GMT
server: uvicorn
content-disposition: attachment; filename="My_History.zip"
x-archive-files: zip
x-frame-options: SAMEORIGIN
x-request-id: 009fc8c0099240cb90738a75a40fa39f
Transfer-Encoding: chunked
Connection: close

- 2 /home/davelopez/dev/galaxy/database/objects/9/e/6/dataset_9e64d76e-a142-4e05-8f39-d849dfa0abbc.dat My_History/1.Dataset_A.txt
- 2 /home/davelopez/dev/galaxy/database/objects/a/4/1/dataset_a41cb8e5-29c9-4eba-ac40-8f29a5af1aad.dat My_History/2.Dataset_B.txt
- 2 /home/davelopez/dev/galaxy/database/objects/0/6/4/dataset_06435695-3c6e-4af6-9513-2d7c291c6320.dat My_History/3.Dataset_C.txt
```

I hope this is ok now :)

mvdbeek replied:
Yep, looks like in the [docs](https://github.com/evanmiller/mod_zip#usage).

mvdbeek reviewed (commented):
This looks really good, I think my only concern is with breaking mod_zip for zip archive creation. I'll see if I can figure out what's up with the integration tests.

davelopez committed:
Apply suggestions from code review

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Update HistoryContentsArchiveDryRunResult docstring

davelopez committed:
Fix `archive` operation to support `upstream_mod_zip`

davelopez committed:
Add fallback to legacy url_for in UrlBuilder

davelopez committed:
Add names to routes for generating url matching

davelopez committed:
Override URL serialization using UrlBuilder

davelopez committed:
Fix merge error

The type_id should be optional

davelopez committed:
Fix `contents_url` generation for collections

davelopez committed:
Add missing alternative update route

davelopez committed:
Reorder routes to avoid masking

The order of the routes declaration is important. The route with the most specific path must be declared before the more general one.

davelopez committed:
Fix response model issues

Enabling this flag will avoid returning fields with default values.

The `type_id` field was not returned by the services dictionary response but the pydantic model was adding it with the default value of None. This was causing the client to behave in a strange way when editing the name of a collection, making a POST request to create a new collection instead of making a PUT request with the name change.

davelopez committed:
Inject UrlBuilder when serializing dataset collection instances

davelopez committed:
Remove test case for datetime format not supported by `datetime.fromisoformat`

davelopez committed:
Force explicit JSON payload in tests

When deleting a dataset through the history contents API

davelopez committed:
Adapt contents_near endpoint to use new direction parameter

davelopez committed:
Adapt changes from #12914

davelopez committed:
Fix mypy

davelopez committed:
Remove duplicated code

davelopez committed:
Remove raw API decorators from legacy endpoints

The return pydantic models/dicts now, not raw JSON

mvdbeek reviewed (approved):
That is amazing, thanks so much @davelopez!
